File: 28. Top K closest points to origin.cpp
/*
 Task: Find the top K closest points to origin (2D plane)
*/

#include <bits/stdc++.h>
using namespace std;

vector<pair<int, int>> kClosest(vector<pair<int, int>>& pts, int k)
{
    priority_queue<pair<int, pair<int, int>>> pq;  // max-heap by dist
    for (auto& p : pts)
    {
        int d = p.first * p.first + p.second * p.second;
        pq.push({d, p});
        if ((int)pq.size() > k)
            pq.pop();
    }
    vector<pair<int, int>> res;
    while (!pq.empty())
    {
        res.push_back(pq.top().second);
        pq.pop();
    }
    return res;
}

int main()
{
    vector<pair<int, int>> pts = {{1, 3}, {-2, 2}, {5, 8}, {0, 1}};
    auto r = kClosest(pts, 2);
    for (auto& p : r) cout << "(" << p.first << "," << p.second << ") ";
    cout << "\n";
    return 0;
}

// Time Complexity: O(n log k)
// Space Complexity: O(k)

File: 28. Top K closest points to origin.py
"""
Task: Find the top K closest points to origin (2D plane)
Time Complexity: O(n log k)
Space Complexity: O(k)
"""

import heapq

def k_closest(pts,k):
    pq=[]
    for x,y in pts:
        d = x*x + y*y
        heapq.heappush(pq, (-d, x, y))
        if len(pq)>k: heapq.heappop(pq)
    return [(x,y) for (_,x,y) in pq]

print(k_closest([(1,3),(-2,2),(5,8),(0,1)],2))
